### ISSUE TITLE
Tanner crashes after dirb scan without error

### ISSUE THREAD

**Issue by t3chn0m4g3** [mushorg/tanner#353]
This issue was originally posted in the Snare repository (https://github.com/mushorg/snare/issues/233), however I could break it down to tanner (ver. 0.6.0, tcp/8090).

At some point while running `dirb http://1.2.3.4` Snare looses contact to Tanner and throws an error 
```
2020-02-17 13:23:01 ERROR:aiohttp.server:log_exception: Error handling request
2020-02-17 13:28:42 ERROR:snare.tanner_handler:submit_data: Exception:
Traceback (most recent call last):
File "/usr/lib/python3.7/site-packages/Snare-0.3.0-py3.7.egg/snare/tanner_handler.py", line 53, in submit_data
timeout=10.0
File "/usr/lib/python3.7/site-packages/aiohttp/client.py", line 391, in _request
await resp.start(conn)
File "/usr/lib/python3.7/site-packages/aiohttp/client_reqrep.py", line 770, in start
self._continue = None
File "/usr/lib/python3.7/site-packages/aiohttp/helpers.py", line 673, in exit
raise asyncio.TimeoutError from None
concurrent.futures._base.TimeoutError
2020-02-17 13:28:42 ERROR:aiohttp.server:log_exception: Error handling request
```
The reason for the error lies within the fact that Tanner crashes without any error or warning message.

Setup follows `docker-compose.yml` of this repo (https://github.com/mushorg/tanner/blob/master/docker/docker-compose.yml).

**Comment by afeena** [comment]
Thanks @t3chn0m4g3! I will investigate this behavior in the next few days.

**Comment by AndreyMironenkoGit** [comment]
Problem in KEEP_ALIVE_TIME = 75 in (tanner/tanner/session.py string 13) and subsequent removal of old sessions from redis (async def delete_old_sessions(self, redis_client) in tanner/tanner/session_manager.py string 67). The temporary mitigation may be to increase the KEEP_ALIVE_TIME, but need to understand why there is a conflict in Redis. I can't understand because I'm not a programmer and i have paws 😊

**Comment by AndreyMironenkoGit** [comment]
Checked the redis, is ok. Problem in tanner/tanner/session_manager.py in sting 71 (is_deleted = await self.delete_session(sess, redis_client)). "await" does not work correctly.

**Comment by afeena** [comment]
@AndreyMironenkoGit are you willing to make a pull request?

**Comment by AndreyMironenkoGit** [comment]
I'm sorry, but no. I don't have a fork. And I don't plan on doing it.

**Comment by mzfr** [comment]
@afeena I've made the PR with the changes suggested by @AndreyMironenkoGit

**Comment by Fosocles** [comment]
@mzfr I recompiled a docker image with your fix and it still crashes after using Nikto against the page Snare serves.

**Comment by afeena** [comment]
OK I am now diving deep into the problem. 
Unfortunately I don't think changing specified Exception to general one make a trick. The more import is to understand why it is happening and fix the cause

**Comment by AndreyMironenkoGit** [comment]
@Fosocles try it exactly as I wrote it. I was testing from 3 computers running 2 Acunetix (full scan) and 1 dirb in parallel. The service was working fine. The test lasted 20 hours.

**Comment by afeena** [comment]
@AndreyMironenkoGit am I correct, that you delete completely this line from your code?
`is_deleted = await self.delete_session(sess, redis_client)`?

**Comment by Fosocles** [comment]
@AndreyMironenkoGit tried both versions and they both fail. Am I doing something wrong?

**Comment by AndreyMironenkoGit** [comment]
@afeena Yep. And did print exception in stdout, and run dirb.

**Comment by afeena** [comment]
@Fosocles if you are using pull request #353 it only changes exception type
@AndreyMironenkoGit deleted the function of session deletion and analysis completely (actually, it's not how tanner supposed to work :) )

I found the problem. dirb creates thousands of sessions so tanner tries to analyse everything and put into redis, so it's a problem of efficiency and scalability

**Comment by Fosocles** [comment]
The most recent error (from tanner.err) looks like there is a problem with the database:

`2020-02-26 10:16:23 ERROR:aiohttp.server:log_exception: Error handling request
Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/aiohttp/web_protocol.py", line 418, in start
    resp = await task
  File "/usr/lib/python3.7/site-packages/aiohttp/web_app.py", line 458, in _handle
    resp = await handler(request)
  File "/usr/lib/python3.7/site-packages/Tanner-0.6.0-py3.7.egg/tanner/server.py", line 64, in handle_event
    detection = await self.base_handler.handle(data, session)
  File "/usr/lib/python3.7/site-packages/Tanner-0.6.0-py3.7.egg/tanner/emulators/base.py", line 143, in handle
    detection = await self.emulate(data, session)
  File "/usr/lib/python3.7/site-packages/Tanner-0.6.0-py3.7.egg/tanner/emulators/base.py", line 125, in emulate
    detection = await self.handle_get(session, data)
  File "/usr/lib/python3.7/site-packages/Tanner-0.6.0-py3.7.egg/tanner/emulators/base.py", line 101, in handle_get
    possible_get_detection = await self.get_emulation_result(session, get_data, self.get_emulators)
  File "/usr/lib/python3.7/site-packages/Tanner-0.6.0-py3.7.egg/tanner/emulators/base.py", line 73, in get_emulation_result
    emulation_result = await self.emulators[detection['name']].handle(attack_params[detection['name']], session)
  File "/usr/lib/python3.7/site-packages/Tanner-0.6.0-py3.7.egg/tanner/emulators/sqli.py", line 65, in handle
    self.query_map = await self.sqli_emulator.setup_db()
  File "/usr/lib/python3.7/site-packages/Tanner-0.6.0-py3.7.egg/tanner/emulators/sqlite.py", line 20, in setup_db
    await self.helper.setup_db_from_config(self.working_dir, self.db_name)
  File "/usr/lib/python3.7/site-packages/Tanner-0.6.0-py3.7.egg/tanner/utils/sqlite_db_helper.py", line 26, in setup_db_from_config
    conn = sqlite3.connect(db_name)
sqlite3.OperationalError: unable to open database file`

**Comment by AndreyMironenkoGit** [comment]
@afeena in the PR #357  @mzfr function has been removed too.

![image](https://user-images.githubusercontent.com/38126215/75339836-54ddf500-58a2-11ea-88b1-dfcd35174daf.png)

**Comment by afeena** [comment]
@AndreyMironenkoGit my bad :D you are right

**Comment by Fosocles** [comment]
@afeena so what should the fix be?

**Comment by Fosocles** [comment]
Any updates?

**Comment by afeena** [comment]
@Fosocles there is no "fast" solution for the problem.

If we are talking about freezing error with dirb, it happens because dirb requests creates thousands of sessions in TANNER (although it supposed to be one sessions with multiple paths), so now imagine at some point all of them are expired and now TANNER want to analyse them and put into redis --> Yes, it freezes. So the whole process of moving sessions from memory to redis/other storage should be re-written. Because now TANNER just goes through the whole list of sessions at once

**Comment by Fosocles** [comment]
And what should I do? I'm trying to integrate Snare/Tanner in my Master's Thesis and I need to be able to collect relevant data with it. What could I do to "fix" the freeze while I'm not actively monitoring Tanner?

**Comment by afeena** [comment]
Please test https://github.com/mushorg/tanner/pull/362

Now doesn't crash for me with dirb

**Comment by afeena** [comment]
#362 
For any other problems please create new issue
